File: 653.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def findTarget(self, root, k):
        """
        :type root: TreeNode
        :type k: int
        :rtype: bool
        """
        if not root:
            return False
        stack, set1 = [root], set()
        while stack:
            node = stack.pop()
            if k - node.val in set1:
                return True
            set1.add(node.val)
            if node.left: stack.append(node.left)
            if node.right: stack.append(node.right)
        return False

File: 653.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool findTarget(TreeNode* root, int k) {
        vector<int> res;
        stack<TreeNode*> q;
        while(!q.empty() or root != NULL){
            if (root != NULL){
                q.push(root);
                root = root->left;
            }else{
                TreeNode* node = q.top();
                q.pop();
                res.push_back(node->val);
                root = node->right;
            }
        }
        unordered_map<int, int> umap;
        for (int i = 0; i < res.size(); i++){
            int tmp = k - res[i];
            if (umap.find(tmp) != umap.end()){
                return true;//{umap.find(tmp) -> second, i};
            }
            umap.insert(pair<int, int>(res[i],i));
        }
        return false;
    }
};
